cli/internal/runcache/runcache.go

```diff
@@ -164,17 +170,33 @@ func (tc TaskCache) RestoreOutputs(terminal *cli.PrefixedUi, logger hclog.Logger
 		}
 		return false, nil
 	}
-	// TODO(gsoltis): check if we need to restore goes here
-	// That will be an opportunity to prune down the set of outputs as well
-	hit, _, _, err := tc.rc.cache.Fetch(tc.rc.repoRoot.ToString(), tc.hash, tc.repoRelativeGlobs)
+	changedOutputGlobs, err := tc.rc.outputWatcher.GetChangedOutputs(tc.hash, tc.repoRelativeGlobs)
 	if err != nil {
-		return false, err
-	} else if !hit {
-		if tc.rc.cacheMissLogsMode != NoLogs {
-			terminal.Output(fmt.Sprintf("cache miss, executing %s", ui.Dim(tc.hash)))
+		logger.Warn(fmt.Sprintf("Failed to check if we can skip restoring outputs for %v: %v. Proceeding to check cache", tc.pt.TaskID, err))
+		terminal.Warn(ui.Dim(fmt.Sprintf("Failed to check if we can skip restoring outputs for %v: %v. Proceeding to check cache", tc.pt.TaskID, err)))
+		changedOutputGlobs = tc.repoRelativeGlobs
+	}
+	if len(changedOutputGlobs) == 0 {
+		logger.Debug(fmt.Sprintf("Skipping cache check for %v, outputs have not changed since previous run.", tc.pt.TaskID))
+	} else {
+		// Note that we currently don't use the output globs when restoring, but we could in the
+		// future to avoid doing unnecessary file I/O
+		hit, _, _, err := tc.rc.cache.Fetch(tc.rc.repoRoot.ToString(), tc.hash, changedOutputGlobs)
```

Comment:
What does this cause the `OutputWatcher` to do? (Is this an infinite loop of "always invalidated" if you start from empty outputs but a cache hit?)

Reply:
I'm not sure I understand your question. 

If you're asking about the unused `changedOutputGlobs`, it's that both `fsCache.Fetch` and `httpCache.Fetch` don't use that particular value. This is an existing deficiency in our cache abstraction.

If you're asking about what happens with a task with empty outputs, I don't think it can happen. If nothing else, we always include the log file. But supposing we didn't, `len(changedOutputGlobs)` would be `0`, and we skip restoration, which is fine because there's nothing to restore.

Reply:
Sorry, my question wasn't clear. I was wondering what the interaction between cache restoration and `OutputWatcher` would be.

Can cache restoration trigger file watching changes? Do we have a sentinel that says, "hey, this change is coming from our cache, don't retrigger?" It's an edge case that this particular line of code made me think of.

Reply:
Oh, yes, it can and should trigger a filewatch notification. We've either:
 * skipped restoration because the outputs directory has not changed, or
 * we're restoring the outputs from some other task (different hash) and we will need to restore again when we come back to the hash that originally called for the watch on these outputs.